src/multiplexed-session.ts

```diff
@@ -0,0 +1,320 @@
+ * limitations under the License.
+ */
+
+import {EventEmitter} from 'events';
+import PQueue from 'p-queue';
+import {Database} from './database';
+import {Session} from './session';
+import {Transaction} from './transaction';
+import {
+  isDatabaseNotFoundError,
+  isInstanceNotFoundError,
+  isCreateSessionPermissionError,
+  isDefaultCredentialsNotSetError,
+  isProjectIdNotSetInEnvironmentError,
+} from './session-pool';
+
+interface MuxSession {
+  multiplexedSession: Session | null;
+}
+
+/**
+ * @callback GetSessionCallback
+ * @param {?Error} error Request error, if any.
+ * @param {Session} session The read-write session.
+ * @param {Transaction} transaction The transaction object.
+ */
+export interface GetSessionCallback {
+  (
+    err: Error | null,
+    session?: Session | null,
+    transaction?: Transaction | null
+  ): void;
+}
+
+/**
+ * Interface for implementing multiplexed session logic, it should extend the
+ * {@link https://nodejs.org/api/events.html|EventEmitter} class
+ *
+ * @interface MultiplexedSessionInterface
+ * @extends external:{@link https://nodejs.org/api/events.html|EventEmitter}
+ *
+ * @constructs MultiplexedSessionInterface
+ * @param {Database} database The database to create a multiplexed session for.
+ */
+export interface MultiplexedSessionInterface {
+  /**
+   * When called creates a multiplexed session.
+   *
+   * @name MultiplexedSessionInterface#createSession
+   */
+  createSession(): void;
+
+  /**
+   * When called returns a multiplexed session.
+   *
+   * @name MultiplexedSessionInterface#getSession
+   * @param {GetSessionCallback} callback The callback function.
+   */
+  getSession(callback: GetSessionCallback): void;
+}
+
+/**
+ * Multiplexed session configuration options.
+ *
+ * @typedef {object} MultiplexedSessionOptions
+ * @property {number} [refreshRate=7] How often to check for expiration of multiplexed session, in
+ *     days. Must not be greater than 7 days.
+ */
+
+export interface MultiplexedSessionOptions {
+  refreshRate?: number;
+  databaseRole?: string | null;
+}
+
+const DEFAULTS: MultiplexedSessionOptions = {
+  refreshRate: 7,
+  databaseRole: null,
+};
+
+/**
+ * Class used to manage connections to Spanner using multiplexed session.
+ *
+ * **You don't need to use this class directly, connections will be handled for
+ * you.**
+ *
+ * @class
+ * @extends {EventEmitter}
+ */
+export class MultiplexedSession
+  extends EventEmitter
+  implements MultiplexedSessionInterface
+{
+  database: Database;
+  options: MultiplexedSessionOptions;
+  _acquires: PQueue;
+  _muxSession!: MuxSession;
+  _pingHandle!: NodeJS.Timer;
+  _refreshHandle!: NodeJS.Timer;
+  constructor(
+    database: Database,
+    multiplexedSessionOptions?: MultiplexedSessionOptions
+  ) {
+    super();
+    this.database = database;
+    this.options = Object.assign({}, DEFAULTS, multiplexedSessionOptions);
+    this.options.databaseRole = this.options.databaseRole
```

Comment:
Where is this `databaseRole` used ? In session pool its used for `batchCreateSession` .  I believe this needs to be passed to `this.database.createSession`

Reply:
`this.database.createSession` is utilizing the database role same as the one passed via pool options, we have
```
if (typeof poolOptions === 'object') {
      this.databaseRole = poolOptions.databaseRole || null;
    }
```
in the database class constructor and this in
```
reqOpts.session.creatorRole =
      options.databaseRole || this.databaseRole || null;
```
      createSession method
      now since we don't have any such thing MultiplexedSessionOptions in other languages and are not exporting MultiplexedSessionOptions to our customers, hence I am removing the MultiplexedSessionOptions from this class as well